### PR TITLE
Fix build authorizer to properly check that user has permissions on site

### PR DESCRIPTION
**Pull request by jseppi**
This site fixes an issue with the build authorizer that would allow _any_ user to create a build model (or restart one). The problem was a buggy if-condition.

Ref #1054

To Do:
- [x] fix tests
- [x] fix logic

### CODE REVIEW

File: test/api/unit/authorizers/build.test.js [hunk 1]
```diff
@@ -1,89 +1,107 @@
...
-      }).then(({ user, build }) => {
-        return authorizer.create(user, { user: user.id, site: site.id })
-      }).then(() => {
-        done(new Error("Expected authorization error"))
-      }).catch(err => {
-        expect(err).to.equal(403)
-        done()
-      }).catch(done)
-    })
+    it("should reject if the build is not associated with one of the user's sites", (done) => {
```
Comment (jseppi) [inline test/api/unit/authorizers/build.test.js]: This is the test whose logic has been amended. It wasn't previously making a site that did belong to the user, so it wasn't really running through the logic of the target method.

File: api/authorizers/build.js [hunk 2]
```diff
@@ -1,25 +1,27 @@
-const { User, Site } = require("../models")
+const { User, Site } = require('../models');
 
-const authorize = (user, build) => {
-  return User.findById(user.id, { include: [Site] }).then(user => {
-    for (site of user.Sites) {
-      if (build.site || build.Site.id === site.id) {
```
Comment (jseppi) [inline api/authorizers/build.js]: This was the problem line. It was resolving the promise just if `build.site` was truthy, but not actually checking to see if the user was associated with the given site.

File: api/authorizers/build.js [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
-        return Promise.resolve()
+const authorize = (user, build) => (
+  User.findById(user.id, { include: [Site] })
+    .then((userModel) => {
+      const buildSiteId = build.site || build.Site.id;
+      const matchingUserSite = userModel.Sites.find(site => buildSiteId === site.id);
```
Comment (jseppi) [inline api/authorizers/build.js]: These two lines are the essential part of th efix

File: api/controllers/build.js [hunk 4]
```diff
@@ -50,14 +50,14 @@ module.exports = {
...
     authorizer.create(req.user, params)
-    .then(() => Build.create(params))
-    .then(build =>
-      GithubBuildStatusReporter.reportBuildStatus(build)
-      .then(() => build)
-    )
-    .then(build => buildSerializer.serialize(build))
-    .then(buildJSON => res.json(buildJSON))
-    .catch(res.error);
+      .then(() => Build.create(params))
```
Comment (jseppi) [inline api/controllers/build.js]: just linting fixes in this file

File: test/api/unit/services/S3PublishedFileLister.test.js [hunk 5]
```diff
@@ -1,143 +1,146 @@
...
-        ])
-        done()
-      }).catch(done)
-    })
-
-    it("should reject with an error if S3.listObjects is unsuccessful", done => {
-      AWSMocks.mocks.S3.listObjects = (params, cb) => cb(new Error("Test error"))
-
-      factory.site().then(model => {
-        return S3PublishedFileLister.listPublishedFilesForBranch(site, "preview")
```
Comment (jseppi) [inline test/api/unit/services/S3PublishedFileLister.test.js]: Interestingly, this test started failing after fixing the build authorizer. If you see here in the old code, `site` was not actually defined! `model` was the name of the argument to the `then(...)` callback, so I'm not totally sure how this test ever worked o_O

File: test/api/unit/services/S3PublishedFileLister.test.js [hunk 6]
```diff
@@ -1,143 +1,146 @@
...
+        done();
+      }).catch(done);
+    });
+
+    it('should reject with an error if S3.listObjects is unsuccessful', (done) => {
+      AWSMocks.mocks.S3.listObjects = (params, cb) => cb(new Error('Test error'));
+
+      factory.site()
+        .then(
+          site => S3PublishedFileLister.listPublishedFilesForBranch(site, 'preview')
```
Comment (jseppi) [inline test/api/unit/services/S3PublishedFileLister.test.js]: Here's the fix to the problem described in the above comment.

**Comment by jseppi** [comment]
ready for review, @jmhooper

File: test/api/unit/authorizers/build.test.js [hunk 7]
```diff
@@ -1,89 +1,107 @@
...
+          done();
+        })
+        .catch(done);
+    });
 
-    it("should reject if the build is not associated with one of the user's sites", done => {
-      Promise.props({
-        user: factory.user(),
-        site: factory.site(),
-      }).then(({ user, build }) => {
```
Comment (jmhooper) [inline test/api/unit/authorizers/build.test.js]: It looks like maybe there was also a typo here in my code? Looks like `build` is not in the original object and `site` is undefined...

File: test/api/unit/services/S3PublishedFileLister.test.js (hunk 5 above)
Comment (jmhooper) [inline test/api/unit/services/S3PublishedFileLister.test.js]: My suspicion would be that maybe somewhere `site = "asdf"` exists without a `var`, `let`, or `const` declaration 😳.

File: api/authorizers/build.js (hunk 2 above)
Comment (jmhooper) [inline api/authorizers/build.js]: Good catch. Good fix. 👍.

**Comment by jseppi** [comment]
Welp, I screwed that up. 🙃

**Comment by jmhooper** [comment]
Did you... delete all the commits?

**Comment by jseppi** [comment]
...yes, but can recover

**Pull request closed without merging**